File: crates/mnemos_daemon/src/events.rs
//! Event bus: broadcasts typed events to all connected WebSocket subscribers.

use serde::{Deserialize, Serialize};
use std::sync::Arc;
use tokio::sync::broadcast;

/// All events the daemon can emit.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum Event {
    MemoryCreated {
        id: String,
        title: String,
        tier: String,
    },
    MemoryUpdated {
        id: String,
    },
    MemoryInvalidated {
        id: String,
        reason: Option<String>,
    },
    SessionStarted {
        id: String,
    },
    SessionEnded {
        id: String,
    },
    ChunkAdded {
        session_id: String,
        chunk_id: String,
    },
    PipelineCompleted {
        session_id: String,
        facts_added: usize,
    },
    PipelineFailed {
        session_id: String,
        error: String,
    },
    ReflectionCompleted {
        reflections_created: usize,
    },
    CommunityDetected {
        communities: usize,
    },
    SyncStarted {
        backend: String,
        direction: String,
    },
    SyncCompleted {
        backend: String,
        direction: String,
        files_changed: usize,
    },
    SyncFailed {
        backend: String,
        direction: String,
        error: String,
    },
    SyncConflict {
        path: String,
        detected_by: String,
    },
    EmbedRebuildStarted {
        target_kind: String,
        target_model: String,
        target_dim: u32,
    },
    EmbedRebuildProgress {
        processed: usize,
        total: usize,
    },
    EmbedRebuildCompleted {
        processed: usize,
        skipped: usize,
        total: usize,
    },
    EmbedRebuildFailed {
        error: String,
        processed: usize,
    },
}

/// Cloneable handle to the broadcast channel that backs all WebSocket subscriptions.
#[derive(Clone)]
pub struct EventBus {
    tx: Arc<broadcast::Sender<Event>>,
}

impl EventBus {
    pub fn new() -> Self {
        let (tx, _) = broadcast::channel(256);
        Self { tx: Arc::new(tx) }
    }

    /// Returns a new receiver that will see all future events.
    pub fn subscribe(&self) -> broadcast::Receiver<Event> {
        self.tx.subscribe()
    }

    /// Fire-and-forget publish. Silently drops if there are no subscribers.
    pub fn publish(&self, e: Event) {
        let _ = self.tx.send(e);
    }
}

impl Default for EventBus {
    fn default() -> Self {
        Self::new()
    }
}

File: desktop/src/store/events.ts
import { create } from "zustand";

export type DaemonEvent =
  | { type: "memory_created"; id: string; title: string; tier: string }
  | { type: "memory_updated"; id: string }
  | { type: "memory_invalidated"; id: string; reason: string | null }
  | { type: "session_started"; id: string }
  | { type: "session_ended"; id: string }
  | { type: "pipeline_completed"; session_id: string; facts_added: number }
  | { type: "pipeline_failed"; session_id: string; error: string }
  | { type: "reflection_completed"; reflections_created: number }
  | { type: "community_detected"; communities: number }
  | { type: "sync_started"; backend: string; direction: string }
  | { type: "sync_completed"; backend: string; direction: string; files_changed: number }
  | { type: "sync_failed"; backend: string; direction: string; error: string }
  | { type: "sync_conflict"; path: string; detected_by: string }
  // P2-15: embed_rebuild_* variants so exhaustive narrowing over DaemonEvent
  // stays correct and ws.ts INVALIDATE table keys resolve to real event types.
  | { type: "embed_rebuild_started"; target_kind: string; target_model: string }
  | { type: "embed_rebuild_progress"; processed: number; total: number }
  | {
      type: "embed_rebuild_completed";
      processed: number;
      skipped: number;
      total: number;
      swapped: boolean;
    }
  | { type: "embed_rebuild_failed"; error: string; processed: number }
  | { type: "backfill_started"; total: number }
  | { type: "backfill_progress"; processed: number; total: number; entities_linked: number; errors: number }
  | { type: "backfill_completed"; total: number; entities_linked: number; edges_created: number; errors: number };

type Status = "connecting" | "open" | "closed";

interface EventState {
  status: Status;
  recent: DaemonEvent[];
  setStatus: (s: Status) => void;
  push: (e: DaemonEvent) => void;
}

export const useEventStore = create<EventState>((set) => ({
  status: "connecting",
  recent: [],
  setStatus: (status) => set({ status }),
  push: (e) => set((st) => ({ recent: [e, ...st.recent].slice(0, 50) })),
}));
